### PR TITLE
src/cunumeric: handle high number of bins in GPU bincount

### LINKED ISSUE #503: Handle high #bins in bincount GPU variant
When I try to run this program on a Tesla V100:

```
x = cn.array([0, 20000], dtype=np.int64)
reps = cn.bincount(x)
```

I hit an assertion inside `bincount.cu`:

```
legion_python: cunumeric/stat/bincount.cu:174: void cunumeric::BincountImplBody<cunumeric::VariantKind::GPU, CODE>::operator()(legate::AccessorRD<Legion::SumReduction<long int>, false, 1>, legate::AccessorRO<typename legate::LegateTypeOf<CODE>::type, 1>&, Legion::Rect<1>&, Legion::Rect<1>&) const [with legate_core_type_code_t CODE = INT64_LT; legate::AccessorRD<Legion::SumReduction<long int>, false, 1> = Legion::ReductionAccessor<Legion::SumReduction<long int>, false, 1, long long int, Realm::AffineAccessor<long int, 1, long long int>, false>; legate::AccessorRO<typename legate::LegateTypeOf<CODE>::type, 1> = Legion::FieldAccessor<LEGION_READ_PRIV, long int, 1, long long int, Realm::AffineAccessor<long int, 1, long long int>, false>; typename legate::LegateTypeOf<CODE>::type = long int; Legion::Rect<1> = Realm::Rect<1, long long int>]: Assertion `num_ctas > 0' failed.
```

Apparently for #bins close to 20k there is not enough memory to instantiate even one threadblock, as returned by `cudaOccupancyMaxActiveBlocksPerMultiprocessor`:

```
num_ctas = 0 THREADS_PER_BLOCK = 128 num_bins = 20001 bin_size = 80004
```

We could do multiple passes over the input array, as allowed by the CUDA occupancy limits, to fill in consecutive slices of the bins array.

At the very least this error should be passed to the python side as a cunumeric exception.

### PR DESCRIPTION
**Pull request by rohany**
The existing bincount implementation on GPUs attempts to allocate
a workspace for all bins within the shared memory available on each
SM. This commit updates the implementation to fall back to a slower
kernel that reduces to global memory when there are too many bins to
fit into shared memory.

Fixes #503.

### CODE REVIEW

File: tests/integration/test_bincount.py [hunk 1]
```diff
@@ -48,6 +49,30 @@ def test_bincount_weights(dtype):
...
+@pytest.mark.parametrize("dtype", DTYPES)
+def test_bincount_weights_high_bins(dtype):
+    v_num = num.array([0, LARGE_NUM_BINS], dtype=dtype)
+    w_num = num.random.randn(2)
+
+    v_np = v_num.__array__()
+    w_np = w_num.__array__()
+
+    out_np = np.bincount(v_np, weights=w_np)
+    out_num = num.bincount(v_num, weights=w_num)
```
Comment (bryevdv) [inline tests/integration/test_bincount.py:72]:
```suggestion
    v_num = num.array([0, LARGE_NUM_BINS], dtype=dtype)
    w_num = num.random.randn(2)
    out_num = num.bincount(v_num, weights=w_num)
    
    v_np = v_num.__array__()
    w_np = w_num.__array__()
    out_np = np.bincount(v_np, weights=w_np)
```
This is is a very minor suggestion, but a grouping like this reads nicely as:

* prepare cn value
* prepare np value
* compare cn and np values

**Comment by bryevdv** [comment]
@rohany I don't have any comment on the cuda code but the Python tests look 👍

File: tests/integration/test_bincount.py (hunk 1 above)
Comment (rohany) [reply tests/integration/test_bincount.py]: Done, cleaned up the other tests in the file as well.

File: src/cunumeric/stat/bincount.cu [hunk 2]
```diff
@@ -98,6 +98,20 @@ static __global__ void __launch_bounds__(THREADS_PER_BLOCK, MIN_CTAS_PER_SM)
   }
 }
 
+template <typename VAL>
+static __global__ void bincount_kernel_rd_global(AccessorRD<SumReduction<int64_t>, false, 1> lhs,
+                                                 AccessorRO<VAL, 1> rhs,
+                                                 const size_t volume,
+                                                 const size_t num_bins,
```
Comment (magnatelee) [inline src/cunumeric/stat/bincount.cu:105]: `num_bins` is unused in the kernel

**Comment by magnatelee** [comment]
@rohany left a minor comment. looks good to me otherwise. we will wait for #546 to go first.

one comment: an alternative I can think of is to subdivide the bins into chunks, each of which fits to the shared memory, and make multiple kernel launches. this would require scanning the input multiple times, but might be more efficient in case the bin size is huge. any thoughts?

**Comment by manopapad** [comment]
> an alternative I can think of is to subdivide the bins into chunks, each of which fits to the shared memory, and make multiple kernel launches. this would require scanning the input multiple times, but might be more efficient in case the bin size is huge.

That was the other alternative proposed in the original issue https://github.com/nv-legate/cunumeric/issues/503. I assume @rohany went with this solution because it was more straightforward.

I can imagine cases where different alternatives would be the better choice, so ideally we would have both implementations, and select dynamically based on some heuristic.

However, in the grand scheme of things I would expect the case of 20k+ bins to be pretty rare in practice, so I don't think we need to put too much effort in this case (but feel free to disagree).

File: src/cunumeric/stat/bincount.cu (hunk 2 above)
Comment (rohany) [reply src/cunumeric/stat/bincount.cu]: Removed in most recent version.

**Comment by rohany** [comment]
Yes, you're right @manopapad. I spoke to @magnatelee offline, and I'll merge this and convert the existing issue to look into exploring different options here.

**Pull request merged**